File: 22FALL/498.diagonal-traverse.py
#
# @lc app=leetcode id=498 lang=python3
#
# [498] Diagonal Traverse
#

# @lc code=start
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        m, n = len(mat), len(mat[0])

        res = []
        left_to_right = True

        for ij_sum in range(0, m+n-1):
            if left_to_right:
                i = min(ij_sum, m-1)
                j = ij_sum-i
            else:
                j = min(ij_sum, n-1)
                i = ij_sum-j

            while 0 <= i <= min(m-1, ij_sum) and (0 <= j <= min(n-1, ij_sum)):
                res.append(mat[i][j])
                if left_to_right:
                    i -= 1
                    j += 1
                else:
                    i += 1
                    j -= 1
            left_to_right = not left_to_right
        
        return res

                

# @lc code=end

File: VSCode_CPP/498.diagonal-traverse.cpp
/*
 * @lc app=leetcode id=498 lang=cpp
 *
 * [498] Diagonal Traverse
 */

// @lc code=start
class Solution {
   public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        vector<int> res;
        if (mat.empty() || mat[0].empty()) return res;
        int m = mat.size(), n = mat[0].size();
        for (int dia = 0; dia < m + n - 1; dia++) {
            if (dia % 2 == 0) {
                for (int x = min(dia, m - 1); x >= 0 && dia - x < n; x--)
                    res.push_back(mat[x][dia - x]);
            } else {
                for (int x = max(0, dia - n + 1); dia - x >= 0 && x < m; x++)
                    res.push_back(mat[x][dia - x]);
            }
        }
        return res;
    }
};
// @lc code=end
